File: Data Structures & Algorithms/binary-tree-diameter/submission-15.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) { 

        const res = [0];

        this.dfs(root, res); 

        return res[0]
    } 
     
    dfs(node, res) {
        if (!node) return 0;
        const l = this.dfs(node.left, res);
        const r = this.dfs(node.right, res);
        res[0] = Math.max(res[0], l + r);
        return 1 + Math.max(l , r);
    } 

}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-15.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
         
        def dfs(curNode, left, right): 
            if not curNode:
                return True 

            if left < curNode.val < right: 
                return dfs(curNode.left, left, curNode.val) and dfs(curNode.right, curNode.val, right) 

            return False 

        return dfs(root, float('-inf'), float('inf'))